### ISSUE TITLE
gradle/gradle should work when imported into IntelliJ

### ISSUE THREAD

**Issue by eskatos** [gradle/gradle#3980]
instead of using `./gradlew idea`.

### Expected Behavior
All IDE features that work when using `./gradlew idea` also work when importing the project into IntelliJ.

Not using "One module per source set".

### Current Behavior

See [./gradle/idea.gradle](https://github.com/gradle/gradle/blob/master/gradle/idea.gradle) for current setup.

*Blockers*
- [ ] Run configuration for remote debugging
   https://youtrack.jetbrains.com/issue/IDEA-174087
- [ ] Code style configuration (can be shared in project import mode)
   https://youtrack.jetbrains.com/issue/IDEA-174087
- [ ] Compilers memory settings
   https://youtrack.jetbrains.com/issue/IDEA-174087
- [ ] Import
  - [ ] Project depending on project with higher language level fails the import (`testing-junit-platform`)
      One work around is to set all projects language level to 8 for IntelliJ, like we do today with 6.
- [ ] Build project
  - [ ] Project dependencies to non-default configurations ignored (test fixtures)
      https://youtrack.jetbrains.com/issue/IDEA-183838
      __This is the main blocker.__
- [ ] Run integration tests
  - [ ] Default JUnit run configuration for integration tests
     https://youtrack.jetbrains.com/issue/IDEA-174087
     Cannot use the same trick as for the remote debugging configuration above because most of the parameters are dynamically created by the build.
  - [ ] The `gradle/gradle` integration tests not running when importing the build.
      https://github.com/gradle/gradle/pull/4451 
   

*Nice to have*
- [ ] Disable IntelliJ suggestion to add web/android flavors to the project
   https://youtrack.jetbrains.com/issue/IDEA-174087
- [ ] Tasks automatically run on project import
   https://youtrack.jetbrains.com/issue/IDEA-176770

Also see #2560

**Comment by lptr** [comment]
As soon as we use the imported project, all compilation will be delegated to Gradle. Since we don't have incremental Groovy compilation support in Gradle today, this would make working with tests very painful. Implementing incremental Groovy compilation is not a requirement for the import itself to function, but to do any meaningful work on the codebase would be close to impossible without it.

**Comment by eskatos** [comment]
@lptr I agree our ultimate goal should be to enable delegation to Gradle but the intent of this issue is to make good first step by just importing. Please open another issue about delegation.

**Comment by eskatos** [comment]
Since c22f324128ee2d530c3de66543551cda12edfe20, project import using one module per project doesn't work anymore :/

**Comment by adammurdoch** [comment]
Most of the blockers in the description look like they would go away if we delegated to Gradle.

I also wouldn't include incremental Groovy compile as a blocker. I run tests from the command-line pretty regularly (but much more often from the IDE) and the performance is fine. In fact, I notice IDEA's groovy compile time much more often, as it tends to invalidate and recompile everything pretty regularly.

**Comment by lptr** [comment]
Can we measure IntelliJ vs non-incremental Groovy compiler somehow? I'm afraid to delegate to a non-incremental compiler, but maybe that fear is unfounded, and we could already get benefits by delegating to Gradle, even without doing any of #6062.

**Comment by ldaley** [comment]
>  I run tests from the command-line pretty regularly (but much more often from the IDE) and the performance is fine

This definitely isn't the case on another one of our projects. On that one, I would not accept having to pay the cost of a full recompile of the test code for every iterative change to a test.

**Comment by adammurdoch** [comment]
We should try it and see, for the `gradle/gradle` build, otherwise we're just speculating. Personally, the improvements I get from delegating to Gradle are worth it to me and the sooner I can do so, the better. I haven't found Gradle's Groovy compile on the `gradle/gradle` build to be a problem for my day-to-day work, and IDEA's Groovy compile suffers from various serious performance issues that affect me frequently. This certainly might not be the case for other builds, but those aren't in scope for this particular issue.

**Comment by adammurdoch** [comment]
Ah yes, we should add that one to the list of things that to get fixed. IDEA should just delegate to the existing tooling API test execution APIs to run tests rather than making these assumptions. In the meantime I suspect we could find an (ugly) workaround for the `gradle/gradle` build to get the delegation to work.

This is probably a blocker (at least for me) as I tend to run integration tests more often than unit tests.

**Comment by lptr** [comment]
Missing features like this are exactly why we should fix the import. :)

BTW, even if we cannot enable delegation, importing to IntelliJ alone is already worth it. Not having to deal with `NoClassDefFoundError` because of changed dependencies whenever I switch back and forth between `master` and my branch (or having to run `./gradlew idea` and invalidating all of IntelliJ's caches) is already a big step forward.

**Comment by nskvortsov** [comment]
Also, requirements in this issue are covered by [an extension to gradle idea settings DSL](https://github.com/jetbrains/gradle-idea-ext-plugin). After importing the project, one is able to run integration and functional tests using IDEA's incremental compiler.
As far as I know, some fixes in Kotlin are required, probably delivered in 1.2.60.  @eskatos , can you confirm?

**Comment by eskatos** [comment]
@nskvortsov exactly, this is blocked by the upgrade to Kotlin 1.2.60. We are currently integrating a version of `gradle/kotlin-dsl` that embeds Kotlin 1.2.60-eap-44 into `gradle/gradle`. Work on this issue will be resumed once this is done.

**Comment by nskvortsov** [comment]
Here is [an issue in IDEA tracker](https://youtrack.jetbrains.com/issue/IDEA-169795) that describes wrong test task delegation. Please vote for it, so we can give it correct priority.

**Comment by eskatos** [comment]
@nskvortsov youtrack says one cannot vote on a resolved issue

**Comment by nskvortsov** [comment]
@eskatos, sorry. Reopened.

**Comment by wolfs** [comment]
A PR trying to fix the import: #6523

**Comment by wolfs** [comment]
I opened an issue to support the remaining code style settings we use in `gradle/gradle`: https://github.com/JetBrains/gradle-idea-ext-plugin/issues/40

**Comment by wolfs** [comment]
This is done. For current problems see https://github.com/gradle/gradle/issues/6941.
